fuse_models/include/fuse_models/unicycle_2d_state_cost_functor.h

```diff
@@ -147,28 +143,39 @@ bool Unicycle2DStateCostFunctor::operator()(
   const T* const acc_linear2,
   T* residual) const
 {
-  T position_pred[2];
-  T yaw_pred[1];
+  T delta_position_pred[2];
+  T delta_yaw_pred[1];
   T vel_linear_pred[2];
   T vel_yaw_pred[1];
   T acc_linear_pred[2];
+  T init_position[2];
+  T init_yaw[1];
+  init_position[0] = T(0.0);
+  init_position[1] = T(0.0);
+  init_yaw[0] = T(0.0);
   predict(
-    position1,
-    yaw1,
+    init_position,
+    init_yaw,
     vel_linear1,
     vel_yaw1,
     acc_linear1,
     T(dt_),
-    position_pred,
-    yaw_pred,
+    delta_position_pred,
+    delta_yaw_pred,
     vel_linear_pred,
     vel_yaw_pred,
     acc_linear_pred);
 
+  const T delta_yaw_est = yaw2[0] - yaw1[0];
+  const Eigen::Matrix<T, 2, 1> position1_map(position1[0], position1[1]);
+  const Eigen::Matrix<T, 2, 1> position2_map(position2[0], position2[1]);
+  const Eigen::Matrix<T, 2, 1> position_diff = (position2_map - position1_map);
+  const fuse_core::Matrix<T, 2, 2> R_yaw_inv = fuse_core::rotationMatrix2D(yaw1[0]);
+  Eigen::Map<Eigen::Matrix<T, 2, 1>> delta_position_pred_map(delta_position_pred);
+
   Eigen::Map<Eigen::Matrix<T, 8, 1>> residuals_map(residual);
-  residuals_map(0) = position2[0] - position_pred[0];
-  residuals_map(1) = position2[1] - position_pred[1];
-  residuals_map(2) = yaw2[0] - yaw_pred[0];
+  residuals_map.block(0, 0, 2, 1) = R_yaw_inv * position_diff - delta_position_pred_map;
```

Comment:
I believe a fixed-size block expression is more efficient https://eigen.tuxfamily.org/dox/group__TutorialBlockOperations.html
```suggestion
  residuals_map.block<2, 1>(0, 0) = R_yaw_inv * position_diff - delta_position_pred_map;
```

Reply:
this cant be used for templated arguments

Reply:
I guess it'd have to be
```suggestion
  residuals_map.template block<2, 1>(0, 0) = R_yaw_inv * position_diff - delta_position_pred_map;
```

:point_right: But now that I look at the actual `residuals_map` type, I think you can even do
```suggestion
  residuals_map.template head<2>() = R_yaw_inv * position_diff - delta_position_pred_map;
```
or, if that doesn't work, the dynamic-size expression
```suggestion
  residuals_map.head(2) = R_yaw_inv * position_diff - delta_position_pred_map;
```